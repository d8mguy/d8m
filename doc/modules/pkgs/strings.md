## Strings

This module provides access to a subset of the functions in the corresponding go package, plus a few others.


#### function split

    val split = \(strg, sep: string) -> list(string) {...}
Be careful: go `Split` doesn't return an empty string if no instances of `sep` in `strg`, it returns a singleton list of strg. Thus, the correct test for no separators is split.count == 1, not 0.

#### function join

    val join = \(strgs: list(string), inter: string) -> string {...}


#### function toLower

    val toLower = \(strg: string) -> string {...}

#### function toUpper

    val toUpper = \(strg: string) -> string {...}

#### function toCap

    val toCap = \(strg: string) -> string {...}

#### function toCamel

    val toCamel = \(strgs: list(string), initial: boolean) -> string {...}
Generate a camel case string from the input strgs; `initial` says whether to capitalize the first word.

#### function genUpper

    val genUpper = \imp(inx: integer) -> string {...}
`Inx` encodes a base-26 number; generate an upper case letter sequence from it.

#### function genLower

    val genLower = \imp(inx: integer) -> string {...}
`Inx` encodes a base-26 number; generate a lower case letter sequence from it.

#### function toOrdinal

    val toOrdinal = \imp(xxx: integer) -> string {...}
This isn't the only way to generate ordinals but it's reasonable.

#### function startsWith

    val startsWith = \(strg, ptn: string) -> boolean {...}
True if `strg` starts with `ptn`.

#### function endsWith

    val endsWith = \(strg, ptn: string) -> boolean {...}
True if `strg` ends with `ptn`.

#### function trimSpace

    val trimSpace = \imp(strg: string) -> string {...}
Trim whitespace from both ends. This is done in a go pkg so we use that.

