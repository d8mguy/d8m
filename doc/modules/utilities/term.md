## Term

Module Term provides a `Term` type suitable for parser outputs, and which is used by the lexparse module.

#### Term:type

    val Term = extend tuple(kind: string, arg: ortype(terms: list(Term), strg: string), loc:integer) with {
        method stringify : \() -> string
        method asString = \() -> string
        method asTerms = \() -> list(Term)
        method child0 = \() -> Term
        method child1 = \() -> Term
        method child2 = \() -> Term
        method childN = \(n:integer) -> Term
        method asLeaf = \() -> string       // get the value of a leaf node
        method hasKind = \(wanted:string) -> boolean {...}
        method hasMatch = \(matchfn: \(Term)->boolean) -> boolean {...}
    }

The `asString` and `asTerms` methods interpret `self.arg` as a specific type. These methods are normally applied based on the value of `kind`; if lexparse creates the Term, `kind` is determined by the production being reduced. The four `child` methods interpret self.arg as a list(Term) and return the child Term at the indicated index in the list; `asLeaf` treats child 0 as a string. The `hasKind` and `hasMatch` methods traverse self to determine whether a given kind occurs (for `hasKind`) or the `matchfn` matches on any child element of self.

Access to the Term's `loc` is intended to be accessed directly through the attribute, and is for recording of the location in a parsed string where a Term is found. Since Term isn't provided with a lithook method you must provide a `loc` in the litform creating each Term.
