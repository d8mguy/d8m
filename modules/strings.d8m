/*
 * This module provides access to a subset of the functions in the corresponding go pkg, plus a few others.
 */
import go "strings" as gs

export split, join, toLower, toUpper, toCap, toCamel, genUpper, genLower, toOrdinal, startsWith, endsWith, trimSpace

val join = \(strgs: list(string), inter: string) -> string { gs.Join(strgs, inter) }
// Be careful: gs.Split doesn't return an empty string if no instances of sep in strg, it returns a singleton list of strg
// Thus, the correct test for no separators is split.count = 1, not 0.
val split = \(strg, sep: string) -> list(string) { gs.Split(strg, sep) }
val toLower = \(strg: string) -> string { gs.ToLower(strg) }
val toUpper = \(strg: string) -> string { gs.ToUpper(strg) }

val toCap = \imp(strg: string) -> string {
    var ret: list(byte) = strg
    if(strg != "") {
        val byt0 = strg[0]
        if('a' <= byt0 && byt0 <= 'z') ret[0] = 'A' + byt0 - 'a'
    }
    cvt(ret, string)
}

// Generate a camel case string from the input strgs; initial says whether to capitalize the first word.
val toCamel = \(strgs: list(string), initial: boolean) -> string {
    if(initial) join(strgs.{this.toCap}, "")
    else if(strgs.count == 1) strgs[0]
    else if(strgs.count == 0) ""
    else strgs[0] + join(tail(strgs).{this.toCap}, "")
}

// Generate letter sequence elements from indices for various underlying sets of letters
val genUpper = \imp(inx: integer) -> string {
    val letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    var inx0 = inx
    var bytes: list(byte) = []
    while(inx0 > 0) { pushf(bytes, letters[inx0 % 26]); inx0 /= 26 }
    cvt(bytes, string)
}

val genLower = \imp(inx: integer) {
    val letters = "abdefghijklmnopqrstuvwxyz"
    var inx0 = inx
    var bytes: list(byte) = []
    while(inx0 > 0) { pushf(bytes, letters[inx0 % 26]); inx0 /= 26 }
    cvt(bytes, string)
}

// This isn't the only way to generate ordinals but it's reasonable.
val toOrdinal = \imp(xxx: integer) -> string {
    var x0 = xxx
    var prefix = ""
    if(x0 < 0) { x0 = -x0; prefix = "-" }
    var suffix = "th"
    if(x0 % 10 == 1 && x0 != 11) suffix = "st"
    else if(x0 % 10 == 2 && x0 != 12) suffix = "nd"
    prefix + "#{x0}" + suffix
}

// true if strg starts with ptn
val startsWith = \(strg, ptn: string) -> boolean {
    val cnt = strg.count
    if(ptn.count <= cnt) strg[0...cnt-ptn.count] == ptn
    else false
}

// true if strg ends with ptn
val endsWith = \(strg, ptn: string) -> boolean {
    val cnt = strg.count
    if(ptn.count <= cnt) strg[cnt-ptn.count...cnt] == ptn
    else false
}

// trim whitespace from both ends. This is done in a go pkg so we use that
val trimSpace = \imp(strg: string) -> string {
    gs.TrimSpace(strg)
}

