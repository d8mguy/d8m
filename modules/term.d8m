
import "strings"

export Term

// A flexible version of Term and some methods to use it. Minimalist: just three attributes. The strg form of arg
// is for "leaf" nodes; the terms form is for internal (non-leaf) nodes. The loc slot is optional but can help
// convey info about where (presumably in a text file) the term comes from. Lexparse produces Terms.
// We define some helper methods like asString, asTerms, childN etc. Also a couple of recursive helpers: hasKind
// is a predicate on existence of a matching kind value anywhere in a Term; hasMatch looks for fn-based matches.
val Term: type = extend tuple(kind: string, arg: ortype(terms: list(Term), strg: string), loc:integer) where {
    // Note a subtlety: this is recursive so the rettype must be explicit
    method stringify = \() -> string {
        val args = (tag(arg) == :strg ? arg : strings.join(arg.{stringify(this)}, ";"))
        "[#{kind}: #{args}]"
    }
    method asString = \() { always(x~self.arg, tag(x) == :strg) }
    method asTerms = \() { always(x~self.arg, tag(x) == :terms) }
    method child0 = \() { asTerms(self)[0] }
    method child1 = \() { asTerms(self)[1] }
    method child2 = \() { asTerms(self)[2] }
    method childN = \(n:integer) { asTerms(self)[n] }
    method asLeaf = \() { asString(self.child0) }       // get the value of a leaf node
    method hasKind = \(wanted:string) -> boolean {
        if(tag(arg) == :strg) kind == wanted else any(arg.{ this.hasKind(wanted) })
    }
    method hasMatch = \(matchfn: \(Term)->boolean) -> boolean {
        if(matchfn(self)) true
        else if(tag(arg) == :strg) false
        else any(arg.{ this.hasMatch(matchfn)})
    }
}
