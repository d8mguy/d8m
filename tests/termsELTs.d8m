// termsELTs: test multiple ELTs named types and various Term related code fragments
import "term" melted
import "strings"
val actions = label(:must, :optional, :visit, :start, :starttkn, :reduce, :suppress)
val wgBaseTypes = label(:integer, :float, :string, :boolean, :label, :submitter, :any, :error)
val vpair = extend tuple(tkn: string, stateinx: integer) where {
    method stringify = \() { "V[#{tkn}: #{stateinx}]"}
}
val actionItem = extend tuple(action: actions, v: ortype(m: string, inx: integer, v: list(vpair))) where {
    method asString = \() { always(x~self.v, tag(x) == :m) }
    method asIndex = \() { always(x~self.v, tag(x) == :inx) }
    method asVisits = \() { always(x~self.v, tag(x) == :v) }
    method stringify = \() {
        val innards = case {
            action == :must || action == :start || action == :starttkn || action == :reduce || action == :suppress => self.asString
            action == :optional => stringify(self.asIndex)
            action == :visit => strings.join(self.asVisits.{stringify(this)}, ",")
        }
        "[#{action}:#{innards}]"
    }
}
val string2WGBase = \(s:string) -> wgBaseTypes {   // string to ELT translation fn for wgBaseTypes
    val inx = ["integer", "float", "string", "boolean", "label", "submitter", "any"][this == s => index]
    if(inx == nil) :error else cast(inx, wgBaseTypes)
}
val wgType = extend tuple(leaf:wgBaseTypes, lists:integer) where {
    method lithook = \mod(trm: Term) {
        lists = 0
        var cur = trm
        while(cur.kind == "listx") { cur = cur.child0; lists += 1 }
        leaf = string2WGBase(asString(cur))
    }
}
val bdgpt = extend tuple(ident: string, typ:wgType) where {
    method lithook = \mod(trm: Term) {
        ident = asString(trm.child0)
        typ = [wgType: trm.child1]
    }
}
val foo = \imp() {
    val trm0 = [Term: "colonpair", [[Term: "ident", "yyy", 0], [Term: "listx", [[Term: "ident", "label", 0]], 0]], 0]
    val bp1 = [bdgpt: trm0]
    val ai1 = [actionItem: :optional, 12]
    println(ai1, bp1)
}
foo()
