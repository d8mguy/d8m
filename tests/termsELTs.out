[1:12] [yyy; [4; 1]]
